src/cli/self_update.rs

```diff
@@ -853,6 +853,30 @@ fn install_proxies_with_opts(process: &Process, force_hard_links: bool) -> Resul
     Ok(())
 }
 
+fn check_proxy_sanity(process: &Process, components: &[&str]) -> Result<()> {
+    let bin_path = process.cargo_home()?.join("bin");
+
+    // Sometimes linking a proxy produces an unpredictable result, where the proxy
+    // is in place, but manages to not call rustup correctly. One way to make sure we
+    // don't run headfirst into the wall is to at least try and run our freshly
+    // installed proxies, to see if they return some manner of reasonable output.
+    for component in components {
+        let cmd = Command::new(bin_path.join(format!("{}{EXE_SUFFIX}", component)))
+            .args(["--version"])
+            .status();
+
+        if let Ok(status) = cmd {
```

Comment:
Nit: `if !cmd.is_ok_and(|status| status.success())` might be better in terms of deduplication.